### PR TITLE
feature: Support multiple registered GraphQLContext contributors, allowing them to provide GraphQLContext key/value entries

### PR DESCRIPTION
**Pull request by theJC**
Changes in this PR
----
Implementation to add new feature as described in #1006 -- provide a mechanism to be able to register context contributors whom are able to add to the GraphQLContext and have access to the WebRequest/ServerRequest

Issue #1006

### CODE REVIEW

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import graphql.GraphQLContext
+import org.springframework.http.HttpHeaders
+import org.springframework.web.context.request.WebRequest
+
+interface GraphQLContextContributor {
+    fun contribute(
+        builder: GraphQLContext.Builder,
+        extensions: Map<String, Any>?,
+        headers: HttpHeaders?,
+        webRequest: WebRequest?
```
Comment (theJC) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:28]: Anything else we should make available via this signature?
ie:
-         variables: Map<String, Any>,
-         extensions: Map<String, Any>?,
-         operationName: String?,

Comment (berngp) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:28]: Aren't the headers available in the webrequest?

**Comment by berngp** [comment]
I am wondering if we should invoke the context contributors when we know the query is valid. I.e an instrumentation will invoke the contributors when query validation has passed.

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt (hunk 1 above)
Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:32]: Was just thinking that, that what I get from copying DgsCustomContextBuilderWithRequest ;)

Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:32]: Hmm.... I tried but then hit friction when looking at DefaultDgsQueryExecutor.getJsonResult ... that passes in headers but null for the webRequest....

Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:32]: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DefaultDgsQueryExecutor.kt#L154-L155

Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:32]: Looks like I just need to add an additional signature executeAndExtractJsonPath on interface DgsQueryExecutor and its implementation

Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:30]: Done.

Comment (theJC) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:30]: Anything else we should make available via this signature?
ie:

query: String
variables: Map<String, Any>
operationName: String?

**Comment by theJC** [comment]
> I am wondering if we should invoke the context contributors when we know the query is valid. I.e an instrumentation will invoke the contributors when query validation has passed.

Query parsing consults the context for ParsingOptions.

One could anticipate that a GraphQLContext contributor may want to set parser options based on headers/caller/etc....

See: https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/ParseAndValidate.java#L58-L62

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * For each bean implementing this interface found, the framework will call the [build] method for every request.
+ * The [contribute] method is then able to use the [GraphQLContext.Builder] to provide additional entries to place in the context.
+ */
+interface GraphQLContextContributor {
+    fun contribute(
+        builder: GraphQLContext.Builder,
+        extensions: Map<String, Any>?,
+        webRequest: WebRequest?
```
Comment (theJC) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/GraphQLContextContributor.kt:30]: Do you see value by adding additional items to this signature?
ie:

query: String
variables: Map<String, Any>
operationName: String?

File: graphql-dgs-example-java-webflux/src/test/java/ReactiveGraphQLContextContributorTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.http.HttpHeaders;
+import org.springframework.mock.web.reactive.function.server.MockServerRequest;
+import org.springframework.web.reactive.config.WebFluxConfigurationSupport;
+import reactor.core.publisher.Mono;
+
+import static com.netflix.graphql.dgs.example.context.MyContextContributor.CONTEXT_CONTRIBUTOR_HEADER_NAME;
+import static com.netflix.graphql.dgs.example.context.MyContextContributor.CONTEXT_CONTRIBUTOR_HEADER_VALUE;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(classes = {WebFluxConfigurationSupport.class, DgsAutoConfiguration.class, MovieDataFetcher.class, MyContextContributor.class, ExampleInstrumentation.class, DgsWebFluxAutoConfiguration.class, DgsPaginationAutoConfiguration.class})
```
Comment (theJC) [inline graphql-dgs-example-java-webflux/src/test/java/ReactiveGraphQLContextContributorTest.java:36]: See: https://github.com/Netflix/dgs-framework/issues/1066 for why WebFluxConfigurationSupport was added here

**Comment by berngp** [comment]
The more I look at this the more I think we could implement the _context contributors_ as part of an Instrumentation and not as part of the `DgsQueryExecutor`.  The Instrumentation could be created such that it gets a ObjectProvider of _contributors_ and builds an ordered representation of such contributors via the `ObjectProvider`'s `orderedStream` API. The _contributors_ can then use the `@Ordered` annotations to express an order of precedence. I still think that we should invoke the _contributors_ after _validation of the query_ has passed. If we need to change the _Query Parser Options_ I would rather have a specific mechanism to override those, instead of doing it implicitly by a _context contributor_.

**Comment by theJC** [comment]
@berngp - How do you recommend we make available the WebRequest/ServerRequest to said Instrumentation to then be able to be sent into any context contributors?

The primary use case that drove me to develop this capability was so that the context contributor(s) can have access to the headers on the request and based on the existance or value, be able to set something on the GraphQLContext.

**Comment by berngp** [comment]
By the time the first instrumentation is being called the 'DgsContext' is already available vie the Instrumentation Environment. We could extract it from there, will get a poc. This will suggest that if we implement this feature via Instrumentations, we need to ensure the one that calls the context contributors executes before the others.

**Comment by theJC** [comment]
@berngp  -- I've done as you suggested in https://github.com/Netflix/dgs-framework/pull/1145 Please take a look and verify you still feel that is a better direction to move towards, and if so, I'll close this PR in favor of that one.

> By the time the first instrumentation is being called the 'DgsContext' is already available vie the Instrumentation Environment. We could extract it from there, will get a poc. This will suggest that if we implement this feature via Instrumentations, we need to ensure the one that calls the context contributors executes before the others.

**Pull request closed without merging**